#!/bin/sh
patch -p0 < $0
exit 0
Index: compiler/lib/Transforms/IPO/PassManagerBuilder.cpp
===================================================================
--- compiler/lib/Transforms/IPO/PassManagerBuilder.cpp	(LLVM 3.5)
+++ compiler/lib/Transforms/IPO/PassManagerBuilder.cpp	(cppamp-driver-ng-35)
@@ -163,7 +163,12 @@
 
     MPM.add(createInstructionCombiningPass());// Clean up after IPCP & DAE
     addExtensionsToPM(EP_Peephole, MPM);
-    MPM.add(createCFGSimplificationPass());   // Clean up after IPCP & DAE
+    // FIXME: A workaround solution for the tile uniform violation problem of
+    //        Bolt. SimplifyCFGPass uses SimplifyCFG and duplicates
+    //        tile_barrier.wait() which calls amp_barrier. We should provide
+    //        a solution to fix the inheritance of noduplicate attribute
+    //        instead of comment createCFGSimplificationPass.
+    //MPM.add(createCFGSimplificationPass());   // Clean up after IPCP & DAE
   }
 
   // Start of CallGraph SCC passes.
