diff --git a/include/llvm/ADT/Triple.h b/include/llvm/ADT/Triple.h
index c89cf4d..d6565fd 100644
--- a/include/llvm/ADT/Triple.h
+++ b/include/llvm/ADT/Triple.h
@@ -185,7 +185,9 @@ public:
     Cygnus,
     AMDOpenCL,
     CoreCLR,
-    LastEnvironmentType = CoreCLR
+
+    HCC,
+    LastEnvironmentType = HCC
   };
   enum ObjectFormatType {
     UnknownObjectFormat,
diff --git a/lib/Support/Triple.cpp b/lib/Support/Triple.cpp
index 1967fac..b5868b2 100644
--- a/lib/Support/Triple.cpp
+++ b/lib/Support/Triple.cpp
@@ -210,6 +210,7 @@ const char *Triple::getEnvironmentTypeName(EnvironmentType Kind) {
   case Cygnus: return "cygnus";
   case AMDOpenCL: return "amdopencl";
   case CoreCLR: return "coreclr";
+  case HCC: return "hcc";
   }
 
   llvm_unreachable("Invalid EnvironmentType!");
@@ -469,6 +470,7 @@ static Triple::EnvironmentType parseEnvironment(StringRef EnvironmentName) {
     .StartsWith("cygnus", Triple::Cygnus)
     .StartsWith("amdopencl", Triple::AMDOpenCL)
     .StartsWith("coreclr", Triple::CoreCLR)
+    .StartsWith("hcc", Triple::HCC)
     .Default(Triple::UnknownEnvironment);
 }
 
