diff -r 602ecf771bdd src/include/gmac/api.h.in
--- a/src/include/gmac/api.h.in	Sat Sep 29 12:16:20 2012 +0200
+++ b/src/include/gmac/api.h.in	Wed Oct 17 15:10:51 2012 -0500
@@ -274,12 +274,11 @@
 #ifdef __cplusplus
 #include <cassert>
 #include <cstdio>
-#include <cstdlib>
 #else
 #include <assert.h>
 #include <stdio.h>
+#endif
 #include <stdlib.h>
-#endif
 
 static const char *error[] = {
     "No error",
